src/include/session.h

```diff
@@ -52,19 +53,68 @@ typedef TAILQ_HEAD(__wt_cursor_list, __wt_cursor) WT_CURSOR_LIST;
 
 /* Maximum number of buckets to visit during cursor sweep. */
 #define WT_SESSION_CURSOR_SWEEP_MAX 32
+
+/* TODO this name sucks */
+struct __wt_session_metadata {
+    WT_EVENT_HANDLER *event_handler; /* Application's event handlers */
+
+    const char *name; /* Name */
+
+    WT_DATA_HANDLE *dhandle; /* Current data handle */
+
+    struct timespec last_epoch; /* Last epoch time returned */
+
+    WT_ITEM **scratch;     /* Temporary memory for any function */
+    u_int scratch_alloc;   /* Currently allocated */
+    size_t scratch_cached; /* Scratch bytes cached */
+
+    WT_ITEM err; /* Error buffer */
+
+    WT_SESSION_STATS stats;
+
+    /* Sessions have an associated statistics bucket based on its ID. */
+    u_int stat_bucket; /* Statistics bucket offset */
+
+    WT_CONNECTION *connection;
+};
+
 /*
  * WT_SESSION_IMPL --
  *	Implementation of WT_SESSION.
  */
 struct __wt_session_impl {
-    WT_SESSION iface;
-    WT_EVENT_HANDLER *event_handler; /* Application's event handlers */
+    WT_SESSION iface; /* TODO reinstate original comment */
+
+    union {
+        struct {
+
+            WT_EVENT_HANDLER *event_handler; /* Application's event handlers */
+
+            const char *name; /* Name */
+
+            WT_DATA_HANDLE *dhandle; /* Current data handle */
+
+            struct timespec last_epoch; /* Last epoch time returned */
+
+            WT_ITEM **scratch;     /* Temporary memory for any function */
+            u_int scratch_alloc;   /* Currently allocated */
+            size_t scratch_cached; /* Scratch bytes cached */
+
+            WT_ITEM err; /* Error buffer */
+
+            WT_SESSION_STATS stats;
+
+            /* Sessions have an associated statistics bucket based on its ID. */
+            u_int stat_bucket; /* Statistics bucket offset */
+        };
+
+        struct __wt_session_metadata metadata;
+    };
```

Comment:
Just FYI, the weird "union of structs" thing would go away once this is reasonably complete. It's just a hack to let existing code continue working while stuff gets moved out of `__wt_session_impl`.